File: Code-CSharp/Base/Utilities/OptionalEntry.cs
namespace CodeCsharp.Base.Utilities;

public class OptionalEntry {
  public static void OptionalProgram() {
    //initialize variable
    int nonRef = 1;
    //make optional from variable
    int? opt = nonRef;
    //whether optional has value
    bool b1 = opt.HasValue;
    //reset optional
    opt = null;
    //get optional value, or else -1
    int value = opt ?? -1;
    //initialize any type
    object? any = null;
    //whether any has value
    bool b2 = any is null;
    //initialize any type variable
    string str = "my string";
    //make any from any variable
    var any2 = (object)str;
    //build a container to contain any value
    Dictionary<int, object> anyMap = new Dictionary<int, object>();
    //insert values
    anyMap.Add(1,"string1");
    anyMap.Add(2, 3.14);
    anyMap.Add(3, 500);
    //whether any cast successful
    bool b3 = anyMap[2] is double;
    //any cast to optional
    double? anyValue = b3 ? (double)anyMap[2] : null;

  }
}

File: Code-Cpp/code_cpp/base/utilities/optional_entry.cc
#include "optional_entry.h"
#include "../misc/example_object.h"

#include <map>

void OptionalEntry::OptionalProgram() {
  //initialize variable
  int i = 1;
  //make optional from variable
  std::optional<int> opt = std::make_optional(i);
  //whether optional has value
  bool b1 = opt.has_value(); // (boo1)opt;
  //reset optional
  opt = std::nullopt; // opt.reset();
  //get optional value, or else -1
  int value = opt.value_or(-1);
  //initialize any type
  std::any any = std::any();
  //whether any has value
  bool b2 = any.has_value();
  //initialize any type variable
  std::string str = "my string";
  //make any from any variable
  auto any2 = std::make_any<std::string>(str);
  //build a container to contain any value
  std::map<int, std::any> any_map;
  //insert values
  any_map.emplace(1, "string1");
  any_map.emplace(2, 3.14);
  any_map.emplace(3, 500);
  //whether any cast successful
  bool b3 = std::any_cast<double>(&any_map[2]) != nullptr;
  //any cast to optional
  std::optional<double> any_value =
      b3 ? std::make_optional(std::any_cast<double>(any_map[2])) : std::nullopt;
}
bool OptionalEntry::TestAll() {
  bool ret = true;
  if (!ret)
    std::cout << "Optional Test Failed!";
  return ret;
}
